File: WCCBO2/Shizuku2Lib/EnvironmentCommunicator.cs
using System.IO.BACnet;

namespace Shizuku2.BACnet
{

  /// <summary>Shizuku2のEnvironmentモニタとの通信ユーティリティクラス</summary>
  public class EnvironmentCommunicator : DateTimeCommunicator
  {

    #region 定数宣言

    /// <summary>Device ID</summary>
    public const uint ENVIRONMENTMONITOR_DEVICE_ID = 4;

    /// <summary>排他的ポート番号</summary>
    public const int ENVIRONMENTMONITOR_EXCLUSIVE_PORT = 0xBAC0 + (int)ENVIRONMENTMONITOR_DEVICE_ID;

    /// <summary>WeatherモニタのBACnetアドレス</summary>
    private readonly BacnetAddress bacAddress;

    #endregion

    #region 列挙型

    /// <summary>項目</summary>
    private enum MemberNumber
    {
      /// <summary>乾球温度</summary>
      DrybulbTemperature = 1,
      /// <summary>相対湿度</summary>
      RelativeHumdity = 2,
      /// <summary>水平面全天日射</summary>
      GlobalHorizontalRadiation = 3,
      /// <summary>夜間放射</summary>
      NocturnalRadiation = 4,
      /// <summary>合計エネルギー消費量</summary>
      TotalEnergyConsumption = 5,
      /// <summary>平均不満足者率</summary>
      AveragedDissatisfactionRate = 6,
      /// <summary>瞬時エネルギー消費量</summary>
      InstantaneousEnergyConsumption = 7,
      /// <summary>瞬時不満足者率</summary>
      InstantaneousDissatisfactionRate = 8,
    }

    #endregion

    #region コンストラクタ

    /// <summary>インスタンスを初期化する</summary>
    /// <param name="id">通信に使うBACnet DeviceのID</param>
    /// <param name="name">通信に使うBACnet Deviceの名前</param>
    /// <param name="ipAddress">エミュレータのIPアドレス（「xxx.xxx.xxx.xxx」の形式）</param>
    public EnvironmentCommunicator(uint id, string name = "anoymous device", string ipAddress = "127.0.0.1")
      : base(id)
    {
      bacAddress = new BacnetAddress(BacnetAddressTypes.IP, ipAddress + ":" + ENVIRONMENTMONITOR_EXCLUSIVE_PORT.ToString());
    }

    #endregion

    #region 外気条件

    /// <summary>乾球温度[C]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>乾球温度[C]</returns>
    public float GetDrybulbTemperature(out bool succeeded)
    {
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, (uint)MemberNumber.DrybulbTemperature, out succeeded);
    }

    /// <summary>相対湿度[%]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>相対湿度[%]</returns>
    public float GetRelativeHumidity(out bool succeeded)
    {
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, (uint)MemberNumber.RelativeHumdity, out succeeded);
    }

    /// <summary>水平面全天日射[W/m2]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>水平面全天日射[W/m2]</returns>
    public float GetGlobalHorizontalRadiation(out bool succeeded)
    {
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, (uint)MemberNumber.GlobalHorizontalRadiation, out succeeded);
    }

    /// <summary>夜間放射[W/m2]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>夜間放射[W/m2]</returns>
    public float GetNocturnalRadiation(out bool succeeded)
    {
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, (uint)MemberNumber.NocturnalRadiation, out succeeded);
    }

    #endregion

    #region 室内環境

    /// <summary>ゾーン（下部空間）の乾球温度[C]を取得する</summary>
    /// <param name="oUnitIndex">室外機番号（1～4）</param>
    /// <param name="iUnitIndex">室内機番号（1～5）</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>ゾーン（下部空間）の乾球温度[C]</returns>
    public float GetZoneDrybulbTemperature(int oUnitIndex, int iUnitIndex, out bool succeeded)
    {
      uint instNum = (uint)(1000 * oUnitIndex + 100 * iUnitIndex + MemberNumber.DrybulbTemperature);
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    /// <summary>ゾーン（下部空間）の相対湿度[%]を取得する</summary>
    /// <param name="oUnitIndex">室外機番号（1～4）</param>
    /// <param name="iUnitIndex">室内機番号（1～5）</param>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>ゾーン（下部空間）の相対湿度[%]</returns>
    public float GetZoneRelativeHumidity(int oUnitIndex, int iUnitIndex, out bool succeeded)
    {
      uint instNum = (uint)(1000 * oUnitIndex + 100 * iUnitIndex + MemberNumber.RelativeHumdity);
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum, out succeeded);
    }

    #endregion

    #region 成績関連

    /// <summary>合計エネルギー消費量[GJ]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>合計エネルギー消費量[GJ]</returns>
    public float GetTotalEnergyConsumption(out bool succeeded)
    {
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, (uint)MemberNumber.TotalEnergyConsumption, out succeeded);
    }

    /// <summary>平均不満足者率[-]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>平均不満足者率[-]</returns>
    public float GetAveragedDissatisfactionRate(out bool succeeded)
    {
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, (uint)MemberNumber.AveragedDissatisfactionRate, out succeeded);
    }

    /// <summary>瞬時エネルギー消費量[kW]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>瞬時エネルギー消費量[kW]</returns>
    public float GetInstantaneousEnergyConsumption(out bool succeeded)
    {
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, (uint)MemberNumber.InstantaneousEnergyConsumption, out succeeded);
    }

    /// <summary>瞬時不満足者率[-]を取得する</summary>
    /// <param name="succeeded">通信が成功したか否か</param>
    /// <returns>瞬時不満足者率[-]</returns>
    public float GetInstantaneousDissatisfactionRate(out bool succeeded)
    {
      return ReadPresentValue<float>
        (bacAddress, BacnetObjectTypes.OBJECT_ANALOG_INPUT, (uint)MemberNumber.InstantaneousDissatisfactionRate, out succeeded);
    }

    #endregion

    /* 2024.10.01テスト。負荷が大きすぎるために保留
    #region COV登録関連

    private bool covSubscribed = false;

    private bool[,] znTmpCOVSubscribed = new bool[4, 8];
    private double[,] znTmps = new double[4, 8];

    public bool EnableZoneTemperatureCOV(int oUnitIndex, int iUnitIndex)
    {
      //COV未登録の場合は登録
      if (!covSubscribed)
        client.OnCOVNotification += Client_OnCOVNotification;

      //温度がそもそも登録されている場合
      if (znTmpCOVSubscribed[oUnitIndex - 1, iUnitIndex - 1]) return true;

      uint instNum = (uint)(1000 * oUnitIndex + 100 * iUnitIndex + memberNumber.DrybulbTemperature);
      BacnetObjectId boID = new BacnetObjectId(BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum);

      //まず現在値を読み取る
      znTmps[oUnitIndex - 1, iUnitIndex - 1] = GetZoneDrybulbTemperature(oUnitIndex, iUnitIndex, out bool succeeded);
      if (!succeeded) return false;

      //COV登録
      znTmpCOVSubscribed[oUnitIndex - 1, iUnitIndex - 1] = client.SubscribeCOVRequest(bacAddress, boID, instNum, false, false, 3600);
      return znTmpCOVSubscribed[oUnitIndex, iUnitIndex];
    }

    public bool DisableZoneTemperatureCOV(int oUnitIndex, int iUnitIndex)
    {
      //温度がそもそも登録されていない場合
      if (znTmpCOVSubscribed[oUnitIndex - 1, iUnitIndex - 1]) return true;

      uint instNum = (uint)(1000 * oUnitIndex + 100 * iUnitIndex + memberNumber.DrybulbTemperature);
      BacnetObjectId boID = new BacnetObjectId(BacnetObjectTypes.OBJECT_ANALOG_INPUT, instNum);

      //COV解除
      bool rslt = client.SubscribeCOVRequest(bacAddress, boID, instNum, true, false, 3600);
      if(rslt) znTmpCOVSubscribed[oUnitIndex, iUnitIndex] = false;
      return rslt;
    }

    private void Client_OnCOVNotification(
      BacnetClient sender, BacnetAddress adr, byte invokeId, uint subscriberProcessIdentifier, 
      BacnetObjectId initiatingDeviceIdentifier, BacnetObjectId monitoredObjectIdentifier, 
      uint timeRemaining, bool needConfirm, ICollection<BacnetPropertyValue> values, BacnetMaxSegments maxSegments)
    {
      UInt16 port = BitConverter.ToUInt16(new byte[] { adr.adr[5], adr.adr[4] });

      //ポート番号が異なる場合は終了
      if (port != ENVIRONMENTMONITOR_EXCLUSIVE_PORT) return;
      uint instID = monitoredObjectIdentifier.instance;

      //ゾーンの温湿度が変化した場合
      if (
        monitoredObjectIdentifier.type == BacnetObjectTypes.OBJECT_ANALOG_INPUT &&
        1100 < instID && instID < 4500)
      {
        //温度は1の位が1
        if (instID % 10 == 1)
        {
          foreach (BacnetPropertyValue vl in values)
          {
            if (vl.property.propertyIdentifier == 85) //Presentvalue
              znTmps[(instID / 1000) % 10 - 1, (instID / 100) % 10 - 1] = (float)vl.value[0].Value;
          }
          Console.WriteLine(znTmps[(instID / 1000) % 10 - 1, (instID / 100) % 10 - 1]);
        }

        //湿度は1の位が2
        if (instID % 10 == 2)
        {

        }
      }

    }

    
    #endregion
    */
  }
}

File: Samples/python/bacpypes3/EnvironmentCommunicator.py
import time
import asyncio

from enum import Enum

import sys,os
sys.path.append(os.path.dirname(__file__))
import DateTimeCommunicator

class EnvironmentCommunicator(DateTimeCommunicator.DateTimeCommunicator):

# region 定数宣言

    ENVIRONMENTMONITOR_DEVICE_ID = 4

    ENVIRONMENTMONITOR_EXCLUSIVE_PORT = 0xBAC0 + ENVIRONMENTMONITOR_DEVICE_ID

# endregion

# region 列挙型定義

    class _member(Enum):
        # 乾球温度
        DrybulbTemperature=1
        # 相対湿度
        RelativeHumdity=2
        # 水平面全天日射
        GlobalHorizontalRadiation=3
        # 夜間放射
        NocturnalRadiation=4
        # エネルギー消費
        TotalEnergyConsumption=5
        # 不満足者率
        AveragedDissatisfactionRate=6
        # 瞬時エネルギー消費量
        InstantaneousEnergyConsumption=7
        # 瞬時不満足者率
        InstantaneousDissatisfactionRate=8

# endregion

    def __init__(self, id, name='envComm', device_ip='127.0.0.1', emulator_ip='127.0.0.1', time_out_sec=1.0):
        """インスタンスを初期化する

        Args:
            id (int): 通信用のDeviceのID
            name (str): 通信用のDeviceの名前
            device_ip (str): 通信に使うDeviceのIP Address（xxx.xxx.xxx.xxx）
            emulator_ip (str): エミュレータのIP Address（xxx.xxx.xxx.xxx）
            time_out_sec (float): タイムアウトまでの時間[sec]
        """
        super().__init__(id, name, device_ip, emulator_ip, time_out_sec)
        self.target_ip = emulator_ip + ':' + str(self.ENVIRONMENTMONITOR_EXCLUSIVE_PORT)

    async def get_drybulb_temperature(self):
        """外気乾球温度[C]を取得する

        Returns:
            list: 読み取り成功の真偽,外気乾球温度[C]
        """
        return await self.read_present_value(self.target_ip,'analogInput:' + str(self._member.DrybulbTemperature.value))
    
    async def get_relative_humidity(self):
        """外気相対湿度[%]を取得する

        Returns:
            list: 読み取り成功の真偽,外気相対湿度[%]
        """
        return await self.read_present_value(self.target_ip,'analogInput:' + str(self._member.RelativeHumdity.value))
    
    async def get_global_horizontal_radiation(self):
        """水平面全天日射[W/m2]を取得する

        Returns:
            list: 読み取り成功の真偽,水平面全天日射[W/m2]
        """
        return await self.read_present_value(self.target_ip,'analogInput:' + str(self._member.GlobalHorizontalRadiation.value))
    
    async def get_nocturnal_radiation(self):
        """夜間放射[W/m2]を取得する

        Returns:
            list: 読み取り成功の真偽,夜間放射[W/m2]
        """
        return await self.read_present_value(self.target_ip,'analogInput:' + str(self._member.NocturnalRadiation.value))
    
    async def get_total_energy_consumption(self):
        """合計エネルギー消費量[MJ]を取得する

        Returns:
            list: 読み取り成功の真偽,エネルギー消費量[MJ]
        """
        return await self.read_present_value(self.target_ip,'analogInput:' + str(self._member.TotalEnergyConsumption.value))

    async def get_averaged_dissatisfaction_rate(self):
        """平均不満足者率[-]を取得する

        Returns:
            list: 読み取り成功の真偽,平均不満足者率[-]
        """
        return await self.read_present_value(self.target_ip,'analogInput:' + str(self._member.AveragedDissatisfactionRate.value))

    async def get_instantaneous_energy_consumption(self):
        """瞬時エネルギー消費量[kW]を取得する

        Returns:
            list: 読み取り成功の真偽,瞬時エネルギー消費量[kW]
        """
        return await self.read_present_value(self.target_ip,'analogInput:' + str(self._member.InstantaneousEnergyConsumption.value))

    async def get_instantaneous_dissatisfaction_rate(self):
        """瞬時不満足者率[-]を取得する

        Returns:
            list: 読み取り成功の真偽,瞬時不満足者率[-]
        """
        return await self.read_present_value(self.target_ip,'analogInput:' + str(self._member.InstantaneousDissatisfactionRate.value))

    async def get_zone_drybulb_temperature(self,oUnitIndex,iUnitIndex):
        """ゾーン（下部空間）の乾球温度[C]を取得する
        Args:
            oUnitIndex (int): 室外機番号（1～4）
            iUnitIndex (int): 室内機番号（1～5）

        Returns:
            list: 読み取り成功の真偽,ゾーン（下部空間）の乾球温度[C]
        """        
        objNum = 1000 * oUnitIndex + 100 * iUnitIndex + self._member.DrybulbTemperature.value
        return await self.read_present_value(self.target_ip,'analogInput:' + str(objNum))

    async def get_zone_relative_humidity(self,oUnitIndex,iUnitIndex):
        """ゾーン（下部空間）の相対湿度[%]を取得する
        Args:
            oUnitIndex (int): 室外機番号（1～4）
            iUnitIndex (int): 室内機番号（1～5）

        Returns:
            list: 読み取り成功の真偽,ゾーン（下部空間）の相対湿度[%]
        """
        objNum = 1000 * oUnitIndex + 100 * iUnitIndex + self._member.RelativeHumdity.value
        return await self.read_present_value(self.target_ip,'analogInput:' + str(objNum))

async def main():
    wCom = EnvironmentCommunicator(14)

    while True:
        val = await wCom.get_drybulb_temperature()
        print('乾球温度' + ('= ' + '{:.1f}'.format(val[1]) + ' C' if val[0] else ' 通信失敗'))

        val = await wCom.get_relative_humidity()
        print('相対湿度' + ('= ' + '{:.1f}'.format(val[1]) + ' %' if val[0] else ' 通信失敗'))

        val = await wCom.get_global_horizontal_radiation()
        print('水平面全天日射' + ('= ' + '{:.1f}'.format(val[1]) + ' W/m2' if val[0] else ' 通信失敗'))

        val = await wCom.get_nocturnal_radiation()
        print('夜間放射' + ('= ' + '{:.1f}'.format(val[1]) + ' W/m2' if val[0] else ' 通信失敗'))

        val = await wCom.get_zone_drybulb_temperature(2,4)
        print('VRF2-4の温度' + ('= ' + '{:.1f}'.format(val[1]) + ' C' if val[0] else ' 通信失敗'))

        val = await wCom.get_zone_relative_humidity(2,4)
        print('VRF2-4の湿度' + ('= ' + '{:.1f}'.format(val[1]) + ' %' if val[0] else ' 通信失敗'))

        time.sleep(1)

if __name__ == "__main__":
    asyncio.run(main())